indexing-hadoop/src/main/java/io/druid/indexer/InputRowSerde.java

```diff
@@ -149,52 +179,37 @@ public static final InputRow fromBytes(byte[] data, AggregatorFactory[] aggs)
       //Read timestamp
       long timestamp = in.readLong();
 
-      //Read dimensions
-      StringArrayWritable sw = new StringArrayWritable();
-      sw.readFields(in);
-      List<String> dimensions = Arrays.asList(sw.toStrings());
-
-      MapWritable mw = new MapWritable();
-      mw.readFields(in);
-
       Map<String, Object> event = Maps.newHashMap();
 
-      for (String d : dimensions) {
-        Writable v = mw.get(new Text(d));
-
-        if (v == null) {
+      //Read dimensions
+      List<String> dimensions = Lists.newArrayList();
+      int dimNum = WritableUtils.readVInt(in);
+      for(int i = 0; i < dimNum; i++) {
+        String dimension = readString(in);
+        dimensions.add(dimension);
+        List<String> dimensionValues = readStringArray(in);
+        if(dimensionValues==null) {
           continue;
         }
-
-        if (v instanceof Text) {
-          event.put(d, ((Text) v).toString());
-        } else if (v instanceof StringArrayWritable) {
-          event.put(d, Arrays.asList(((StringArrayWritable) v).toStrings()));
+        if(dimensionValues.size() == 1) {
+          event.put(dimension, dimensionValues.get(0));
         } else {
-          throw new ISE("unknown dim value type %s", v.getClass().getName());
+          event.put(dimension, dimensionValues);
         }
       }
 
-      //Read metrics
-      for (AggregatorFactory aggFactory : aggs) {
-        String k = aggFactory.getName();
-        Writable v = mw.get(new Text(k));
-
-        if (v == null) {
-          continue;
-        }
-
-        String t = aggFactory.getTypeName();
-
-        if (t.equals("float")) {
-          event.put(k, ((FloatWritable) v).get());
-        } else if (t.equals("long")) {
-          event.put(k, ((LongWritable) v).get());
+      int metricSize = WritableUtils.readVInt(in);
+      for(int i = 0; i < metricSize; i++) {
+        String metric = readString(in);
+        String type = readString(in);
```

Comment:
type can be encoded as a number instead of string, you can probably use ValueType.XXX.ordinal() as encoding.
io.druid.segment.column.ValueType is an enum defined with the types.